server/authentication_test.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/kv"
+	"github.com/cockroachdb/cockroach/testutils"
+	"github.com/cockroachdb/cockroach/ts"
+)
+
+// Verify client certificate enforcement.
+func TestSSLEnforcement(t *testing.T) {
+	s := StartTestServer(t)
+	defer s.Stop()
+	testCases := []struct {
+		method, key   string
+		certsStatus   int // Status code for https with client certs.
+		noCertsStatus int // Status code for https without client certs.
+	}{
+		// /ui/: basic file server: no auth.
+		{"GET", "/index.html", http.StatusOK, http.StatusOK},
+
+		// /_admin/: server.adminServer: no auth.
+		{"GET", healthPath, http.StatusOK, http.StatusOK},
+
+		// /debug/: server.adminServer: no auth.
+		{"GET", debugEndpoint + "vars", http.StatusOK, http.StatusOK},
+
+		// /_status/: server.statusServer: no auth.
+		{"GET", statusKeyPrefix, http.StatusOK, http.StatusOK},
+
+		// /kv/rest/: kv.RESTServer. Key cannot be empty. Response is NotFound since the key does not exist.
+		{"GET", kv.EntryPrefix + "foo", http.StatusNotFound, http.StatusUnauthorized},
+
+		// /kv/db/: kv.DBServer. These are proto reqs, but we can at least tell get past auth.
+		{"GET", kv.DBPrefix + "Get", http.StatusBadRequest, http.StatusUnauthorized},
+
+		// /ts/: ts.Server.
+		{"GET", ts.URLPrefix, http.StatusNotFound, http.StatusUnauthorized},
+	}
+
+	// HTTPS with client certs.
+	certsContext := testutils.NewTestBaseContext()
+	client, err := certsContext.GetHTTPClient()
+	if err != nil {
+		t.Fatalf("error initializing http client: %s", err)
```

Comment:
somewhat OT: do you find these custom error messages helpful? you already get the line number, seems like `t.Fatal(err)` would be perfectly sufficient.